tensorflow/src/TensorFlow/Internal/FFI.hs

```diff
@@ -122,43 +153,88 @@ shutDownRunner r = do
     -- TODO(gnezdo): manage exceptions better than print.
     either print (const (return ())) =<< waitCatch r
 
-extendGraph :: Raw.Session -> GraphDef -> IO ()
-extendGraph session pb =
-    useProtoAsVoidPtrLen pb $ \ptr len ->
-        checkStatus $ Raw.extendGraph session ptr len
-
+graphImportGraphDef :: Raw.Graph
+                    -> GraphDef
+                    -> (Raw.ImportGraphDefOptions -> IO ())
+                    -> IO ()
+graphImportGraphDef graph pb optionSetter =
+    useProtoAsBuffer pb $ \buffer ->
+        bracket Raw.newImportGraphDefOptions Raw.deleteImportGraphDefOptions $ \importGraphDefOptions -> do
+            optionSetter importGraphDefOptions
+            checkStatus $ Raw.graphImportGraphDef graph buffer importGraphDefOptions
+
+forGraphOperations_ :: Raw.Graph
+                    -> (Raw.Operation -> IO b)
+                    -> IO ()
+forGraphOperations_ graph f = with 0 go
```

Comment:
i think if the graph gets modified during the loop, we'd get unexpected results. can you add a note about that

ideally we'd keep the APIs with sharp edges contained in the `FFI` and `Raw` modules. one idea is to continue exporting an `extendGraph` function from `FFI`, just implement it using the logic you put into `Session`. not a blocker though

Reply:
That sounds like a good idea; I dislike how much from `Internal` has to be exported to make writing the code in `Session` possible.

Reply:
`TF_GraphNextOperation` does not mention anything about modification of the graph; should I add a comment anyway?

Reply:
on second thought, i think we only really append to the graph, so there probably isn't much risk. i think this is fine